src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -566,20 +578,68 @@ internal bool TryRead(out ReadResult result)
             }
         }
 
-        private static void TrySchedule(PipeScheduler scheduler, Action action)
+        private static void TrySchedule<TState>(PipeScheduler scheduler, Action<TState> action, TState state)
         {
             if (action != null)
             {
-                scheduler.Schedule(action);
+                scheduler.Schedule(action, state);
             }
         }
 
-        private static void TrySchedule(PipeScheduler scheduler, Action<object> action, object state)
+        private static void TrySchedule(PipeScheduler scheduler, in CompletionData completionData)
```

Comment:
Why not stick this logic into CompletionData and pass scheduler into PipeAwaitable ctor?

Reply:
I'm not changing the flow as part of this PR. The PipeAwaitable is responsible for returning data and the pipe itself owns the execution. That makes it easy to see the change in place. We can always refactor.

Reply:
Oh wait, I read that wrong. Yeah I can do that.

Reply:
Why do you think that is cleaner? It's just a poco.

Reply:
If `CompletionData` would live in PipeAwaiter then all logic around capturing and scheduling would be in one place.

Reply:
I'm not sure how that will work when the pipe is the one with the synchronization logic. When we call Complete inside of the lock, we store that in a local then schedule outside of the lock. There's probably a way to make it work I just need to think about it a little (or you can if you feel we should make this change).